static/js/ace-editor.ts

```diff
@@ -0,0 +1,337 @@
+    }
+    return exampleEditor;
+  }
+}
+
+export class HedyAceEditor implements HedyEditor {
+  private _editor?: AceAjax.Editor;
+  private _markers?: Markers
+  askPromptOpen: boolean = false;
+
+  /**
+ * Set the highlither rules for a particular level
+ * @param level      
+ */
+  setHighliterForLevel(level: number): void { 
+    const mode = this.getHighlighter(level);
+    this._editor?.session.setMode(mode);
+  }
+
+  /**
+   * @returns the string of the current program in the editor
+   */
+  getValue(): string { 
+    return this._editor!.getValue();
+  }
+
+  /**     
+   * @returns if the editor is set to read-only mode
+   */
+  isReadOnly(): boolean {
+    return this._editor!.getReadOnly();
+  }
+
+  /**
+   * Sets the editor contents.
+   * @param content the content that wants to be set in the editor
+   */
+  setValue(content: string): void {
+    this._editor?.setValue(content, MOVE_CURSOR_TO_END);
+  }
+
+  /**
+   * Trim trailing whitespaces
+   */
+  trimTrailingSpace(): void {
+    try {
+      // This module may or may not exist, so let's be extra careful here.
+      const whitespace = ace.require("ace/ext/whitespace");
+      whitespace.trimTrailingSpace(this._editor!.session, true);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  /**
+   * Resizes the editor after changing its size programatically
+   */
+  resize(): void {
+    this._editor?.resize()
+  }
+
+  /**
+   * Focuses the text area for the current editor
+   */
+  focus(): void { 
+    this._editor?.focus();
+  }
+
+  /**
+   * Clears the errors and annotations in the editor
+   */
+  clearErrors(): void {
+    // Not sure if we use annotations everywhere, but this was
+    // here already.
+    this._editor?.session.clearAnnotations();
+    this.markers?.clearErrors();
+  }
+
+  /**     
+   * Moves to the cursor to the end of the current file
+   */
+  moveCursorToEndOfFile(): void { 
+    this._editor?.navigateFileEnd();
+  }
+
+  /**
+   * Clears the selected text
+   */
+  clearSelection(): void {
+    this._editor?.clearSelection();
+  }
+
+  /**
+  * Removes all breakpoints on the rows.
+  **/
+  clearBreakpoints(): void { 
+    this._editor?.session.clearBreakpoints();
+  }
+
+  /**
+   * If this editor is used as a main editor, we set the options here
+   */
+  configureMainEditor(): void {
+    this._editor?.setShowPrintMargin(false);
+    this._editor?.renderer.setScrollMargin(0, 0, 0, 20)
+    this._editor?.addEventListener('change', () => {
+      theLocalSaveWarning.setProgramLength(this._editor!.getValue().split('\n').length);
+    });
+    // Set const value to determine the current page direction -> useful for ace editor settings
+    const dir = $("body").attr("dir");
+    if (dir === "rtl") {
+      this._editor?.setOptions({ rtl: true });
+    }
+
+    // If prompt is shown and user enters text in the editor, hide the prompt.
+    this._editor?.on('change', () => {
+      if (this.askPromptOpen) {
+        stopit();
```

Comment:
Now the editor needs to know about the JS running, and it needs to know about HTML elements that are not itself.

This is a bit of mixing responsibilities that we probably shouldn't have.

Reply:
I think the best thing to do is to probably make the `HedyEditor` emit a `change` event itself (either by being a full-blown `EventEmitter`, or by making some simple callback mechanism ourselves), and leaving this logic where it was.

We should have some EventEmitter code in a file somewhere. If not, we can always build something like this ourselves

```ts
type ChangeListener = (x: string) => void;

interface HedyEditor {
  public readonly onChange: ChangeListener[];
}

class AceHedyEditor {
  public readonly onChange = new Array<ChangeListener>();

  constructor() {
    this.editor.on('change', () => {
      const contents = this.getContents();
      for (const listener of this.onChange) {
        listener(contents);
      }
    });
  }
}

// Consumer
editor.onChange.push((contents) => {
  console.log('Editor changed to', contents);
});
```

Reply:
I love this idea!